develop-docs/sdk/telemetry/logs.mdx

```diff
@@ -0,0 +1,381 @@
+
+#### JavaScript
+
+```jsx
+// Browser JavaScript - need to rely on tagged template literals for string templating
+Sentry.logger.info(Sentry.logger.fmt`Adding item ${itemId} for user ${userId}`);
+
+// Server-side (Node.js, Bun, Deno) with printf-like syntax
+Sentry.logger.info("Adding item %s for user %s", [itemId, userId], {
+  extra: 123,
+});
+
+Sentry.logger.emit({
+  level: "info",
+  message: "Adding item for user",
+  attributes: {
+    itemId,
+    userId,
+    extra: 123,
+  },
+});
+```
+
+#### Python
+
+```python
+# With f-string like syntax
+Sentry.logger.info('Adding item {item} for user {user}', item=item_id, user=user_id, attributes={ 'extra': 123 });
+
+Sentry.logger.emit(
+  level='info',
+  message='Adding item for user',
+  attributes={
+    'item_id': item_id,
+    'user_id': user_id,
+    extra: 123
+  },
+)
+```
+
+#### PHP
+
+```php
+use function Sentry\logger;
+
+logger()->info('Adding item %s for user %s', [$item_id, $user_id], ['extra' => 123]);
+
+logger()->emit(
+  level: LogLevel::info(),
+  message: 'Adding item for user',
+  attributes: [
+    'item_id': item_id,
+    'user_id': user_id,
+    'extra': 123
+  ],
+);
+```
+
+#### Java
+
+```java
+import io.sentry.Sentry;
+
+// example with MessageFormat based string template
+Sentry.logger().info("Adding item {0,string} for user {1,string}", item_id, user_id);
+
+Sentry.logger().emit(log -> {
+  log.setLevel("info");
+  log.setMessage("Adding item for user");
+  log.setAttribute("item_id", item_id);
+  log.setAttribute("user_id", user_id);
+  log.setAttribute("extra", 123);
+});
+
+// Java - Builder Pattern, where setters return builder for method chaining
+Sentry.Logger.emit(log -> {
+  log.setLevel("info")
+  .setMessage("Adding item for user")
+  .setAttribute("item_id", item_id)
+  .setAttribute("user_id", user_id)
+  .setAttribute("extra", 123)
+});
+
+// Kotlin
+Sentry.Logger.emit("Adding item for user") { // fun emit(block: LogItem.() -> Unit)
+  setLevel("info")
+  setAttribute("item_id", item_id)
+  setAttribute("user_id", user_id)
+  setAttribute("extra", 123)
+}
+```
+
+#### Apple
+
+```swift
+// Swift
+SentrySDK.logger
+  .info(message: "Adding item %s for user %s",
+    params: [itemId, userId],
+    attributes: ["extra": @"123"]
+  )
+
+// Objective-C
+[SentrySDK.logger
+    emit :@"Adding item for user"
+    level: @"info"
+    attributes: @{ @"extra" : @"123", @"item_id" : item_id, @"user_id" : user_id }
+];
+```
+
+#### Threading and Concurrency Considerations
+
+Both the `Sentry.logger.X` and `Sentry.logger.emit` methods are fire-and-forget (have no return value). This means that the SDK can choose to run these methods on a separate thread or queue them up for processing later. This includes evaluating the string template, and running any internal hooks (like `beforeSendLog`). The SDK should ensure that the logs are sent in the order they are received.
```

Comment:
Putting everything immediately on a bg thread has the downside of losing logs that occur directly before a crash. Consider this
```Swift
SentrySDK.logger.warn("Not enough memory")
// Won't see the log above for this crash.
codeThatCrashes()
```

We had this issue in the past with breadcrumbs on Cocoa, and then we made them very efficient so we can store them on the calling thread to have them available for crashes.

If we want to replace breadcrumbs completely with logs, we must support that scenario. Maybe it's OK to skip it for the first iteration. This should impact every mobile SDK. I'm also happy to help figure this out once we start implementing this on mobile. I'm unsure how important that use case is for browser and server.

Reply:
The same applies for server crashes, so we need to figure this out. Left a note about this: 28cb4b597e10634dfff7d645b36094af21c19cc4